Framework/Algorithms/src/XrayAbsorptionCorrection.cpp

```diff
@@ -0,0 +1,234 @@
+  if (detectorAngle > 180) {
+    detectorDistance = -detectorDistance;
+  }
+  Kernel::V3D detectorPos = {x, detectorDistance, 0.};
+  return detectorPos;
+}
+
+/**
+ * normalise moun intensity to 1.
+ * @param muonIntensity A MantidVec which contains the intensity of muons as a
+ * function  of depth
+ * @return A vector of doubles which contains the normalised muon intensity
+ */
+std::vector<double> XrayAbsorptionCorrection::normaliseMuonIntensity(
+    MantidVec const &muonIntensity) {
+  double sum_of_elems{0};
+  for (double elem : muonIntensity) {
+    sum_of_elems += elem;
+  }
+
+  std::vector<double> normalisedMuonIntensity;
+  for (double x : muonIntensity) {
+    normalisedMuonIntensity.push_back(x / sum_of_elems);
+  }
+  return normalisedMuonIntensity;
+}
+
+/**
+ * Calculate the muon implantation position in the sample.
+ * @param muonProfile A reference to the muon profile workspace
+ * @param inputWS A reference to the input workspace
+ * @param detectorDistance a double representing the distance from the sample to
+ * detector
+ * @return A vector of V3D objects that represent the position of muons
+ */
+std::vector<Kernel::V3D> XrayAbsorptionCorrection::calculateMuonPos(
+    API::MatrixWorkspace_sptr &muonProfile, API::MatrixWorkspace_sptr inputWS,
+    double detectorDistance) {
+  const MantidVec muonDepth = muonProfile->readX(0);
+  Kernel::V3D muonPoint = {0., detectorDistance, 0.};
+  Kernel::V3D toStart = {0, -1, 0};
+  const Geometry::IObject *shape = &inputWS->sample().getShape();
+  Geometry::Track muonPath = Geometry::Track(muonPoint, toStart);
+  shape->interceptSurface(muonPath);
+  if (muonPath.count() == 0) {
+    throw std::runtime_error("No valid solution, check shape parameters, Muon "
+                             "depth profile and detector distance");
+  }
+  Kernel::V3D intersection = muonPath.cbegin()->entryPoint;
+  double sampleDepth = intersection[1];
+  std::vector<Kernel::V3D> muonPos;
+  for (auto depth : muonDepth) {
+    Kernel::V3D pos = {0., sampleDepth - (depth / 100), 0.};
+    muonPos.push_back(pos);
+  }
+  return muonPos;
+}
+
+/**
+ * Calculate the mid point of a bin in workspace.
+ * @param energyVector a MantidVec which contains edges of bins
+ * @return A vector of doubles of bin mid points
+ */
+std::vector<double> XrayAbsorptionCorrection::calculateMidPointOfBin(
+    MantidVec const &energyVector) {
+  std::vector<double> midPoint;
+  for (size_t i = 0; i < energyVector.size() - 1; i++) {
+    midPoint.push_back((energyVector[i] + energyVector[i + 1]) / 2);
+  }
+  return midPoint;
+}
+/**
+ * Execution
+ */
+void XrayAbsorptionCorrection::exec() {
+  MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
+  MatrixWorkspace_sptr outputWS = inputWS->clone();
+  MatrixWorkspace_sptr muonProfile = getProperty("MuonImplantationProfile");
+  MantidVec muonIntensity = muonProfile->readY(0);
+  std::vector<double> normalisedMuonIntensity =
```

Comment:
This profile seems to describe the fact that the muon intensity will decrease as you go deeper into the sample. In other neutron absorption correction algorithms this effect is calculated as part of the algorithm rather than being taken as an input.
I guess it's perhaps a bit easier for neutrons because the inbound particle is the same as the outbound particle (ie still a neutron after being scattered). For this kind of instrument you would have to incorporate muon cross section information into Mantid to achieve a calculation of the pre-scatter attenuation. This would be similar to the work you've done to plumb in the X-ray attenuation information I suspect.
Might be a bit late in the day to consider adding this in but thought I'd mention it. Might be worth running past Adrian even if it's a separate piece of work.

Reply:
The Muon position would be calculated before algorithm is called so not included here.

Reply:
@tolu28-coder is that calculation being done with SRIM?

Reply:
yes

Reply:
SRIM is a semi-empirical model used to simulate penetration depths of atoms/ions in a material. It takes into account the ionic and electronic (to some extent) stopping power when calculating the depth. For a negative muon they just use a H^- ion.